src/app/store/courses/courses.reducer.ts

```diff
@@ -1,15 +1,138 @@
-import { Action } from '@ngrx/store';
+import { Action, createReducer, on } from "@ngrx/store";
 
-// Add your code here
+import * as CourseActions from "./courses.actions";
+import { Course } from "@app/models/course.model";
+
+export const coursesFeatureKey = "courses";
 
 export interface CoursesState {
-    // Add your code here
+  allCourses: Course[];
+  course: Course | null;
+  isAllCoursesLoading: boolean;
+  isSingleCourseLoading: boolean;
+  isSearchState: string;
+  errorMessage: string | null;
 }
 
 export const initialState: CoursesState = {
-    // Add your code here
+  allCourses: [],
+  course: null,
+  isAllCoursesLoading: false,
+  isSingleCourseLoading: false,
+  isSearchState: "",
+  errorMessage: null,
 };
 
-export const coursesReducer; // Add your code here
+export const coursesReducer = createReducer(
+  initialState,
+  on(CourseActions.requestAllCourses, (state) => ({
+    ...state,
+    isAllCoursesLoading: true,
+  })),
+  on(CourseActions.requestAllCoursesSuccess, (state, { courses }) => ({
+    ...state,
+    allCourses: courses,
+    isAllCoursesLoading: false,
+  })),
+  on(CourseActions.requestAllCoursesFail, (state, { error }) => ({
+    ...state,
+    errorMessage: error.message,
+    isAllCoursesLoading: false,
+  })),
+  on(CourseActions.requestSingleCourse, (state, { id }) => ({
+    ...state,
+    course: {
+      id,
+      authors: [],
+      creationDate: new Date(),
+      description: "",
+      duration: 0,
+      title: "",
+    },
+    isSingleCourseLoading: true,
+  })),
+  on(CourseActions.requestSingleCourseSuccess, (state, { course }) => ({
+    ...state,
+    course: course,
+    isSingleCourseLoading: false,
+  })),
+  on(CourseActions.requestSingleCourseFail, (state, { error }) => ({
+    ...state,
+    course: null,
+    errorMessage: error.message,
+    isSingleCourseLoading: false,
+  })),
+  on(CourseActions.requestFilteredCourses, (state, { title }) => ({
+    ...state,
+    isSearchState: title,
+    isAllCoursesLoading: true,
+  })),
+  on(CourseActions.requestFilteredCoursesSuccess, (state, { courses }) => ({
+    ...state,
+    allCourses: courses,
+    isAllCoursesLoading: false,
+  })),
+  on(CourseActions.requestFilteredCoursesFail, (state, { error }) => ({
+    ...state,
+    errorMessage: error.message,
+    isAllCoursesLoading: false,
+  })),
+  on(CourseActions.requestDeleteCourse, (state, { id }) => ({
+    ...state,
+    course: {
+      id,
```

Comment:
I think you could also pass id to requestDeleteCourseSuccess instead of adding here a course